plugins/obs-outputs/rtmp-stream.c

```diff
@@ -1065,9 +1098,116 @@ static void check_to_drop_frames(struct rtmp_stream *stream, bool pframes)
+	int current_bitrate = stream->dynamic_bitrate;
+	int initial_bitrate = stream->initial_bitrate;
+
+	/* X264_NAL_HRD_CBR=2 incompatible with dynamic variable bitrate.
+	 * Bitrate is adjusted downwards every second by about 10% for QSV and 20%
+	 * for enc-amf & x264 (tests suggest QSV handles the bitrate changes better).
+	 * Detection threshold of congestion could be chosen larger than 0.15;
+	 * this would mean less reactivity but would allow micro-congestions
+	 * to heal themselves without changing the bitrate.
+	 * We define a minimal bitrate at half the initial one.
+	 */
+	if ( i_nal_hrd != 2 &&	stream->switch_variable_bitrate) {
+
+		if (congestion > 0.15 && current_bitrate > (int)(initial_bitrate / 2) &&
+				(cur_time_ms - last_adjustment_time) > 1000) {
+			current_bitrate = isQSV? (int)(current_bitrate / 1.1):(int)(current_bitrate / 1.2);
+
+			if (current_bitrate < (int)(initial_bitrate / 2)) {
+				current_bitrate = (int)(initial_bitrate / 2);
+			}
+
+			stream->dynamic_bitrate = current_bitrate;
+			int previous_bitrate = obs_data_get_int(params, "bitrate");
+			obs_data_set_int(params, "bitrate", current_bitrate);
+			obs_encoder_update(vencoder, params);
+			blog(LOG_INFO, "Congestion detected %f percent," 
+					"dropping bitrate for encoder %s "
+					"from %i kbps  to %i kbps", congestion * 100, encoder_id,
+					previous_bitrate, current_bitrate);
+			stream->last_adjustment_time = cur_time_ms;
+		}
+		/* bitrate is adjusted upwards by 10% every 5 sec. This means it will take at most
+		 * 25 - 30 sec to recover initial bitrate once congestion is cleared
+		 * this interval could be decreased but this would mean more oscillations
+		 * and also more packet drops.
+		 * So we choose to have lower quality instead of better with more drops.
+		 */
+		if (mean_congestion < 0.05 && max_congestion < 0.15 &&
+				congestion < 0.05 && last_congestion < 0.05 &&
+				current_bitrate < initial_bitrate &&
+				(cur_time_ms - last_adjustment_time) > 5000) {
+
+			current_bitrate = (int)(current_bitrate * 1.10);
+
+			if (current_bitrate > initial_bitrate) {
+				current_bitrate = initial_bitrate;
+			}
+
+			stream->dynamic_bitrate = current_bitrate;
+			int previous_bitrate = obs_data_get_int(params, "bitrate");
+			obs_data_set_int(params, "bitrate", current_bitrate);
+			obs_encoder_update(vencoder, params);
+			blog(LOG_INFO, "Congestion clearing at %f percent,"
+					"raising bitrate for encoder %s"
+					"to %i kbps from previous bitrate %i kbps \n",
+					congestion * 100, encoder_id, current_bitrate,
+					previous_bitrate);
+			stream->last_adjustment_time = cur_time_ms;
+		}
+		stream->last_congestion = congestion;
+	}
+
+}
+
 static bool add_video_packet(struct rtmp_stream *stream,
 		struct encoder_packet *packet)
 {
+	obs_encoder_t *vencoder = obs_output_get_video_encoder(stream->output);
+	const char *encoder_id = obs_encoder_get_id(vencoder);
+	if (strcmp(encoder_id, "ffmpeg_nvenc") != 0) {
```

Comment:
I think for future additions of more encoders it would be better to not hard code this exclusion but instead let the encoder specify if it supports dynamic bitrate changes or not.

Reply:
that's a good point.   
That was the lazy option I took.    
It would require to add a flag to h264 encoders.   
Will do that as well as other improvements; I forgot to check videotoolbox for instance.